File: logic/src/main/java/by/rekuts/travelagency/logic/repository/Specification.java
package by.rekuts.travelagency.logic.repository;

import javax.persistence.criteria.CriteriaBuilder;
import javax.persistence.criteria.Predicate;
import javax.persistence.criteria.Root;
import java.util.List;

public interface Specification<T> {
    List<Predicate> getPredicates(Root<T> root, CriteriaBuilder builder);
}

File: repository/src/main/scala/by/rekuts/travelagency/repository/Specification.scala
package by.rekuts.travelagency.repository

import java.util

import javax.persistence.criteria.{CriteriaBuilder, Predicate, Root}

trait Specification[T] {

  var id: Int = _

  def getPredicates(root: Root[T], builder: CriteriaBuilder): util.List[Predicate] = {
    var predicates = new util.ArrayList[Predicate]
    if (id != 0) {
      predicates.add(
        builder.equal(root.get("id"), id)
      )
    }
    predicates
  }
}
